File: visao-computacional/src/mqtt_service.py
"""Cliente MQTT para publicar eventos no broker HiveMQ."""
import paho.mqtt.client as mqtt

from . import config

CLIENT_ID = "seceye-python-cv"


class MQTTService:
    def __init__(self):
        self.broker = config.MQTT_BROKER
        self.port = config.MQTT_PORT
        self.topic = config.MQTT_TOPIC
        self.client_id = CLIENT_ID
        self.connected = False
        self.client = mqtt.Client(client_id=self.client_id, clean_session=True)
        self.client.on_connect = self._on_connect
        self.client.on_disconnect = self._on_disconnect
        self.client.reconnect_delay_set(min_delay=1, max_delay=30)

    def _on_connect(self, client, userdata, flags, rc):
        if rc == 0:
            self.connected = True
            print(f">> [MQTT] Conectado ao broker {self.broker}:{self.port}", flush=True)
        else:
            self.connected = False
            print(f">> [MQTT] Falha na conexao (rc={rc})", flush=True)

    def _on_disconnect(self, client, userdata, rc):
        self.connected = False
        print(f">> [MQTT] Desconectado (rc={rc})", flush=True)

    def connect(self):
        try:
            self.client.connect(self.broker, self.port, keepalive=60)
            self.client.loop_start()
        except Exception as e:
            print(f">> [MQTT] Erro ao conectar: {e}", flush=True)

    def publish_trigger(self, state: str):
        try:
            if not self.connected:
                print(">> [MQTT] Nao conectado - tentando reconectar antes de publicar", flush=True)
                try:
                    self.client.reconnect()
                except Exception as e:
                    print(f">> [MQTT] Falha ao reconectar: {e}", flush=True)
                    return

            result = self.client.publish(self.topic, payload=state, qos=1)
            if result.rc == mqtt.MQTT_ERR_SUCCESS:
                print(f">> [MQTT] Publicado '{state}' em {self.topic}", flush=True)
            else:
                print(f">> [MQTT] Falha ao publicar (rc={result.rc})", flush=True)
        except Exception as e:
            print(f">> [MQTT] Erro ao publicar: {e}", flush=True)

    def disconnect(self):
        try:
            self.client.loop_stop()
            self.client.disconnect()
        except Exception as e:
            print(f">> [MQTT] Erro ao desconectar: {e}", flush=True)

File: frontend/src/services/mqtt.service.ts
import mqtt, { MqttClient } from 'mqtt';

const MQTT_BROKER = process.env.EXPO_PUBLIC_MQTT_BROKER;
const MQTT_TOPIC = process.env.EXPO_PUBLIC_MQTT_TOPIC;
const MQTT_USER = process.env.EXPO_PUBLIC_MQTT_USER;
const MQTT_PASS = process.env.EXPO_PUBLIC_MQTT_PASS;

let client: MqttClient | null = null;

function getClient(): MqttClient | null {
  if (client) return client;

  if (!MQTT_BROKER) {
    console.error('[MQTT] EXPO_PUBLIC_MQTT_BROKER nao definido');
    return null;
  }

  try {
    const clientId = 'seceye-app-' + Math.random().toString(16).slice(2, 10);
    const brokerUrl = MQTT_BROKER.trim();

    client = mqtt.connect(brokerUrl, {
      clientId,
      username: MQTT_USER,
      password: MQTT_PASS,
      reconnectPeriod: 3000,
      connectTimeout: 10000,
      clean: true,
    });

    client.on('connect', () => {
      console.log('[MQTT] Conectado ao broker', MQTT_BROKER);
    });

    client.on('reconnect', () => {
      console.log('[MQTT] Reconectando...');
    });

    client.on('error', (err) => {
      console.warn('[MQTT] Erro de conexao:', err?.message ?? err);
    });

    client.on('close', () => {
      console.log('[MQTT] Conexao encerrada');
    });

    client.on('offline', () => {
      console.log('[MQTT] Cliente offline');
    });

    return client;
  } catch (error) {
    console.error('[MQTT] Falha ao inicializar cliente:', error);
    client = null;
    return null;
  }
}

export function connectMqtt(): void {
  try {
    getClient();
  } catch (error) {
    console.error('[MQTT] connectMqtt exception:', error);
  }
}

export function deactivateAlarm(): Promise<void> {
  return new Promise((resolve, reject) => {
    try {
      const c = getClient();

      if (!c) {
        const err = new Error('[MQTT] Cliente nao inicializado');
        console.error(err.message);
        reject(err);
        return;
      }

      if (!MQTT_TOPIC) {
        const err = new Error('[MQTT] EXPO_PUBLIC_MQTT_TOPIC nao definido');
        console.error(err.message);
        reject(err);
        return;
      }

      const doPublish = () => {
        c.publish(MQTT_TOPIC, '0', { qos: 1 }, (err) => {
          if (err) {
            console.error('[MQTT] Falha ao publicar "0":', err.message);
            reject(err);
          } else {
            console.log('[MQTT] Publicado "0" em', MQTT_TOPIC);
            resolve();
          }
        });
      };

      if (c.connected) {
        doPublish();
      } else {
        const timeout = setTimeout(() => {
          c.off('connect', onConnect);
          reject(new Error('[MQTT] Timeout aguardando conexao'));
        }, 10000);

        const onConnect = () => {
          clearTimeout(timeout);
          c.off('connect', onConnect);
          doPublish();
        };

        c.on('connect', onConnect);
      }
    } catch (error) {
      console.error('[MQTT] deactivateAlarm exception:', error);
      reject(error as Error);
    }
  });
}

export function activateAlarm(): Promise<void> {
  return new Promise((resolve, reject) => {
    try {
      const c = getClient();

      if (!c) {
        const err = new Error('[MQTT] Cliente nao inicializado');
        console.error(err.message);
        reject(err);
        return;
      }

      if (!MQTT_TOPIC) {
        const err = new Error('[MQTT] EXPO_PUBLIC_MQTT_TOPIC nao definido');
        console.error(err.message);
        reject(err);
        return;
      }

      const doPublish = () => {
        c.publish(MQTT_TOPIC, '1', { qos: 1 }, (err) => {
          if (err) {
            console.error('[MQTT] Falha ao publicar "1":', err.message);
            reject(err);
          } else {
            console.log('[MQTT] Publicado "1" em', MQTT_TOPIC);
            resolve();
          }
        });
      };

      if (c.connected) {
        doPublish();
      } else {
        const timeout = setTimeout(() => {
          c.off('connect', onConnect);
          reject(new Error('[MQTT] Timeout aguardando conexao'));
        }, 10000);

        const onConnect = () => {
          clearTimeout(timeout);
          c.off('connect', onConnect);
          doPublish();
        };

        c.on('connect', onConnect);
      }
    } catch (error) {
      console.error('[MQTT] activateAlarm exception:', error);
      reject(error as Error);
    }
  });
}

export function disconnectMqtt(): void {
  try {
    if (client) {
      client.end(true, {}, () => {
        console.log('[MQTT] Cliente desconectado');
      });
      client = null;
    }
  } catch (error) {
    console.error('[MQTT] disconnectMqtt exception:', error);
    client = null;
  }
}